sphinx/ext/napoleon/docstring.py

```diff
@@ -872,6 +884,83 @@ def __init__(self, docstring: Union[str, List[str]], config: SphinxConfig = None
         self._directive_sections = ['.. index::']
         super().__init__(docstring, config, app, what, name, obj, options)
 
+    def _convert_type_spec(self, _type):
+        def recombine_set(tokens):
+            def combine_set(tokens):
+                in_set = False
+                set_items = []
+
+                for token in tokens:
+                    if token.startswith("{"):
+                        in_set = True
+                    elif token.endswith("}"):
+                        in_set = False
+                        set_items.append(token)
+
+                    if in_set:
+                        set_items.append(token)
+                    else:
+                        if set_items:
+                            token = "".join(set_items)
+                            set_items = []
+                        yield token
+
+            return list(combine_set(tokens))
+
+        def tokenize_type_spec(spec):
+            delimiters = r"(\sor\s|\sof\s|:\s|,\s|[{]|[}])"
+
+            split = [
+                item
+                for item in re.split(delimiters, _type)
+                if item
+            ]
+            tokens = recombine_set(split)
+            return tokens
+
+        def token_type(token):
+            if token.startswith(" ") or token.endswith(" "):
+                type_ = "delimiter"
+            elif token.startswith("{") and token.endswith("}"):
+                type_ = "value_set"
+            elif token in ("optional", "default"):
+                type_ = "control"
+            elif re.match(":[^:]+:`[^`]+`", token):
+                type_ = "reference"
+            elif token.isnumeric() or (token.startswith('"') and token.endswith('"')):
+                type_ = "literal"
+            else:
+                type_ = "obj"
+
+            return type_
+
+        def convert_obj(obj, translations):
+            return translations.get(obj, ":obj:`{}`".format(obj))
+
+        tokens = tokenize_type_spec(_type)
+        types = [
+            (token, token_type(token))
+            for token in tokens
+        ]
+
+        # TODO: make this configurable
+        translations = {
+            "sequence": ":term:`sequence`",
+            "mapping": ":term:`mapping`",
+            "dict-like": ":term:`dict-like <mapping>`",
+        }
```

Comment:
Why using terms for these entries?

Reply:
these are examples of terms (i.e. not types that could be referenced with `obj`) I have often seen, e.g. in the `pandas` or `xarray` documentation (which is what I'm mostly working with), but I wouldn't put those in the extension. Instead, I'd like to expose a `napoleon_type_aliases` (or whatever) option that allows translating tokens in the type spec to user-specified references.

Reply:
I still don't understand they should be regarded as terms commonly in many documents. So I can accept to provide `napoleon_type_aliases` to change this behavior. And I prefer not to convert them as `:term:` references.

Reply:
I agree, this should not be the default. So I'll try to create the option and remove the `:term:` references.

Reply:
done, but the documentation of the new option could be better